partitioned-heat-conduction/fenics/heat.py

```diff
@@ -160,28 +159,26 @@ def determine_gradient(V_g, u, flux):
 bc = []
 # Create for each dimension of Vbig a coupling expression for either the time derivatives (Dirichlet side)
 #  or Neumann side (no time derivatives required as they are enforced with changing F)
-coupling_expressions = [precice.create_coupling_expression()] * tsm.num_stages
+coupling_expressions = [precice.create_coupling_expression() for _ in range(tsm.num_stages)]
 # for the boundary which is not the coupling boundary, we can just use the boundary conditions as usual
 # each stage needs a boundary condition
 if tsm.num_stages > 1:
     if problem is ProblemType.DIRICHLET:
         for i in range(tsm.num_stages):
             bc.append(DirichletBC(Vbig.sub(i), du_dt[i], remaining_boundary))
-            bc. append(DirichletBC(Vbig.sub(i), coupling_expressions[i], coupling_boundary))
-            # Fixme 4: Damit würde man die richtigen Ergebnisse erhalten
-            # bc.append(DirichletBC(Vbig.sub(i), du_dt[i], coupling_boundary))
+            bc.append(DirichletBC(Vbig.sub(i), coupling_expressions[i], coupling_boundary))
     else:
         vs = split(v)
         for i in range(tsm.num_stages):
             bc.append(DirichletBC(Vbig.sub(i), du_dt[i], remaining_boundary))
-            F += vs[i] * coupling_expressions[i] * dolfin.ds
+            F += vs[i] * coupling_expressions[-1] * dolfin.ds
```

Comment:
It looks like we need to use the heat flux corresponding to the end of the current time step for the Neumann boundary condition. I currently do not have a good explanation why this is the case, but if we think about the way how Neumann boundary conditions work in general and compare this to the way Dirichlet work it intuitively makes sense to me. However, there is certainly a better explanation including more hard facts and less intuition :smirk: 

This problem was masked by the bug in the creation of the coupling expressions because due to the pointer-nature of the problem, all coupling expressions were pointing to the coupling expressions corresponding to the end of the current time step.

Reply:
Another thing we should keep in mind here: If we are using a time stepping scheme, where `c[-1]` does not correspond to the end of the window, things will probably become tricky. Instead of referring to the `coupling_expression` via the index, I think we need a different strategy for creating `coupling_expression`s for the Neumann participant. What's also important: `coupling_expression[0]` is not needed for the Neumann participant. So why bother creating it in the first place?

Reply:
Maybe I was a bit quick here: One thing we should definitely test is what happens, if the heat flux changes over time. Currently, we have a constant heat flux and this might hide some possible errors.

In d410f66abafd372031f036357b96a03f00aa7201 I realized that the heat flux was not initialized (no problem with implicit Euler). This results in zero heat flux for `t=0`, which also leads to an incorrect interpolant. I think this is the much more reasonable fix.

Reply:
> It looks like we need to use the heat flux corresponding to the end of the current time step for the Neumann boundary condition.

Yes, we need to set the heat flux to the end of the time step of the respective stages.
Otherwise it would e.g. not comply with the following statement from the [Irksome paper](https://doi.org/10.48550/arXiv.2006.16282): 

> We note that weakly-enforced boundary conditions [...] require [...] just the evaluation of any time-dependent data at the correct stage times.

So for the weak form for the stage $`k_i`$ at $`t+c_i\cdot\delta t`$.

I tested this as well with $`g_{tri}`$ (from the [Quasi-Newton waveform iteration]( https://doi.org/10.1002/nme.6443) paper) to check if the convergence of the solution is equal to the order of the time stepping scheme. With the introduction of a time dependent heat flux, wrongly imposed Neumann Boundaries would impact the convergence of the solution.
Until now, I tested it with LobattoIIIC(2) and the results look promising:
Neumann BC at $`t+c_i\cdot\delta t`$
| $`\Delta t`$  | $`\delta t`$ | L2 Dirichlet | L2 Neumann|
| ------------- | ------------ | ------------ | ------------- | 
| 0.1  | 0.05 | 1,15E-05 | 7,09E-05|
| 0.05  | 0.025 | 2,59E-06 | 1,74E-05|
| 0.025  | 0.0125 | 6,51E-07| 4,32E-06|

So the convergence order approximately is, as expected, two.

As I was not quite sure, if you referred to the time of each stage by "end of current time step" or the actual end of the current time step, and as a sanity check, if I understood the idea behind the k-ansatz correctly, I also tested, how the errors look like if you only use data at the boundaries retrieved only from $`t+\delta t`$. And, thankfully, the results have no convergence order of two (otherwise I would, again, be clueless why this is happening😅):
| $`\Delta t`$  | $`\delta t`$ | L2 Dirichlet | L2 Neumann|
| ------------- | ------------ | ------------ | ------------- | 
| 0.1  | 0.05 | 2,73E-04 | 3,98E-04
| 0.05  | 0.025 | 1,59E-04 | 2,08E-04
| 0.025  | 0.0125 | 8,96E-05| 1,07E-04

Reply:
Last but not least, a comment about:
>  If we are using a time stepping scheme, where c[-1] does not correspond to the end of the window, things will probably become tricky.

I think this should be fine. As 
> Yes, we need to set the heat flux to the end of the time step of the respective stages. Otherwise it would e.g. not comply with the following statement from the [Irksome paper](https://doi.org/10.48550/arXiv.2006.16282):
> 
> > We note that weakly-enforced boundary conditions [...] require [...] just the evaluation of any time-dependent data at the correct stage times.

Additionally, tested with the [GaussLegendre(1)](https://en.wikipedia.org/wiki/Gauss%E2%80%93Legendre_method), the time dependent heat flux and the Neumann BC only used at $`t+\frac{\delta t}{2}`$, the order seems fine. 
So it is nice to see, that actually no special treatment for such time stepping methods are required. I hope for the best, that this will also be true for multi step methods.